### ISSUE TITLE
Explore the MCMC sampler for a bimodal distribution

### ISSUE THREAD

**Issue by pawel-czyz** [cbg-ethz/PYggdrasil#25]
Consider a distribution $\log p(x) = -\left(x^2-1\right)^2 + \mathrm{const.}$.

Run an MCMC sampler using the BlackJAX library and assess the histogram of samples. How do they compare to the true distribution?

What happens if you wrap the MCMC sampler in the SMC (again, in BlackJAX)?

**Comment by pawel-czyz** [comment]
Time estimates: both M.

**Comment by gordonkoehn** [comment]
Correction:  $ln p(x) =  - (x^2-1)^2$ is meant.

**Comment by pawel-czyz** [comment]
Perhaps useful: https://blackjax-devs.github.io/sampling-book/algorithms/TemperedSMC.html :wink:

**Comment by gordonkoehn** [comment]
Yes, already reading it. ;) Thanks! Just read up on HMC - great post here https://bjlkeng.github.io/posts/hamiltonian-monte-carlo/.

**Comment by gordonkoehn** [comment]
Shared my workings so far in #26 - a PR that is not meant to be merged. Let me know if this is bad practice. Thought this might be useful to share what I have explored so far. Perhaps you might have suggestions on what else to try. 

I have tried to implement a plain Metropolis-Hastings in blackjax - but I somehow don't see how to do it in the blackjax documentation. Is it just a variation of the 'nuts' ?

**Comment by pawel-czyz** [comment]
Nice! As you can see, MCMC algorithms can get stuck on one mode of the distribution.

M–H uses a kernel $q(x_\mathrm{new} \mid x_\mathrm{current})$ to define new samples (which are then accepted or rejected). NUTS and other instances of Hamiltonian Monte Carlo algorithms can be thought as M–H in which the kernel $q$ is neatly constructed to generate high quality samples once the gradient information is available. (However, I've heard that there are some situations in which random walk M–H can outperform HMC, although I haven't ever seen one).

In JAX you can implement M–H with a custom kernel via [rmh](https://blackjax-devs.github.io/blackjax/autoapi/blackjax/kernels/index.html#blackjax.kernels.rmh) – your `proposal` function should be compatible with it (up to the order of function arguments and perhaps a small wrapper to pass from `position` to BlackJAX's `state`, although I'm not sure if it's required here).

**Comment by gordonkoehn** [comment]
Yes, was instructive to see ! 

Thank you for the explanation. - I'll try to create a custom kernel for M-H next then and look into 'nuts' a bit more.

**Comment by gordonkoehn** [comment]
Sufficiently explored behaviour of MCMC for a bimodal target distribution - for now. May reopen/restore to explore SMC.
